### PR TITLE
Score-based generative enhancement model

### PR DESCRIPTION
**Pull request by anteju**
# What does this PR do ?

This PR adds score-based diffusion enhancement model.

**Collection**: ASR

# Changelog 
- Added `PredictiveAudioToAudioModel` class and an example config
- Added `ScoreBasedGenerativeAudioToAudioModel` and an example config
- Added power and scale to `AudioToSpectrogram` and `SpectrogramToAudio`
- Added  diffusion-related blocks (SDEs, samplers, backbone)

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/asr/data/audio_to_audio.py [hunk 1]
```diff
@@ -130,13 +130,23 @@ class ASRAudioProcessor:
         sample_rate: sample rate used for all audio signals
         random_offset: If `True`, offset will be randomized when loading a subsegment
                        from a file.
+        normalize: 
+        normalization_signal:
```
Comment (racoiaws) [inline nemo/collections/asr/data/audio_to_audio.py:134]: Missing argument docs?

File: nemo/collections/asr/data/audio_to_audio.py [hunk 2]
```diff
@@ -314,7 +324,18 @@ def process_audio(self, audio: Dict[str, torch.Tensor]) -> Dict[str, torch.Tenso
         Returns:
             An ordered dictionary of signals and their tensors.
         """
-        # Currently, not doing any processing of the loaded signals.
+        if self.normalize:
+            # Normalize audio w.r.t. to the first signal
```
Comment (racoiaws) [inline nemo/collections/asr/data/audio_to_audio.py:328]: Not clear what "first signal" means

File: nemo/collections/asr/data/audio_to_audio.py [hunk 3]
```diff
@@ -314,7 +324,18 @@ def process_audio(self, audio: Dict[str, torch.Tensor]) -> Dict[str, torch.Tenso
...
+            norm_scale = audio[self.normalization_signal].abs().max()
+
+            # Do not normalize embeddings
+            skip_signals = self.embedding_setup.signals if self.embedding_setup is not None else []
+
+            # Normalize audio signals
+            for signal in audio:
+                if signal not in skip_signals:
+                    audio[signal] = audio[signal] / (norm_scale + self.eps)
```
Comment (racoiaws) [inline nemo/collections/asr/data/audio_to_audio.py:337]: As I understand this code, it assumes that all `signal`s in `audio` are on the same scale to begin with, and we are just rescaling them by the same factor.

Shouldn't normalization depend on some signal statistic? `/= audio[signal] / audio[signal].abs().max() * (norm_scale + self.eps)` or something along these lines?

File: nemo/collections/asr/data/audio_to_audio.py [hunk 4]
```diff
@@ -812,6 +833,7 @@ class AudioToTargetDataset(BaseAudioDataset):
                                 If `None`, all channels will be loaded.
         target_channel_selector: Optional, select subset of channels from each input audio file.
                                  If `None`, all channels will be loaded.
+        normalize_input:
```
Comment (racoiaws) [inline nemo/collections/asr/data/audio_to_audio.py:836]: Missing argument docs?

File: nemo/collections/asr/data/audio_to_audio.py [hunk 5]
```diff
@@ -932,6 +960,7 @@ class AudioToTargetWithReferenceDataset(BaseAudioDataset):
                                    from input and target.
         reference_duration: Optional, can be used to set a fixed duration of the reference utterance. If `None`,
                             complete audio file will be loaded.
+        normalize_input:
```
Comment (racoiaws) [inline nemo/collections/asr/data/audio_to_audio.py:963]: Missing argument docs?

File: nemo/collections/asr/data/audio_to_audio.py [hunk 6]
```diff
@@ -1063,6 +1098,7 @@ class AudioToTargetWithEmbeddingDataset(BaseAudioDataset):
                                 If `None`, all channels will be loaded.
         target_channel_selector: Optional, select subset of channels from each input audio file.
                                  If `None`, all channels will be loaded.
+        normalize_input:
```
Comment (racoiaws) [inline nemo/collections/asr/data/audio_to_audio.py:1101]: Also missing argument docs

File: nemo/collections/asr/data/audio_to_audio.py [hunk 7]
```diff
@@ -1094,7 +1131,12 @@ def __init__(
             max_number=max_utts,
         )
 
-        audio_processor = ASRAudioProcessor(sample_rate=sample_rate, random_offset=random_offset,)
+        audio_processor = ASRAudioProcessor(
+            sample_rate=sample_rate,
+            random_offset=random_offset,
+            normalize=normalize_input,
+            normalization_signal='input_signal',
```
Comment (racoiaws) [inline nemo/collections/asr/data/audio_to_audio.py:1138]: By default, `normalize_input=False`, but we still always pass a string argument for `normalization_signal`.

Would it make sense to condense both into a single `Optional[str]` argument? If it's `None`, then do nothing, if it's something then normalize as specified.

File: nemo/collections/asr/parts/submodules/diffusion.py [hunk 8]
```diff
@@ -0,0 +1,1310 @@
...
+        else:
+            raise RuntimeError(f'Unexpected predictor: {predictor}')
+
+        # Corrector setup
+        if corrector == 'annealed_langevin_dynamics':
+            self.corrector = AnnealedLangevinDynamics(
+                sde=self.sde, score_estimator=score_estimator, snr=snr, num_steps=num_corrector_steps
+            )
+        else:
+            raise RuntimeError(f'Unexpected corrector: {corrector}')
```
Comment (racoiaws) [inline nemo/collections/asr/parts/submodules/diffusion.py:1037]: We have only one option for both predictor and corrector, but list of allowed ones is hardcoded, and we can't pass any external ones. We also can't really subclass `PredictorCorrectorSampler` to do that without fully copying this initializer.

So, it *looks* extensible, there's even OOP boilerplate for a predictor and a corrector. But I see only one ~practical way to extend this with my own code - maintain a fork. This doesn't seem a good approach for a framework.

Do we aim to address this in an initial PR?

File: nemo/collections/asr/models/enhancement_models.py [hunk 9]
```diff
@@ -464,3 +199,380 @@ def list_available_models(cls) -> Optional[PretrainedModelInfo]:
...
+    )
+    def _step(self, target_signal, input_signal, input_length=None):
+        """Randomly generate a time step for each example in the batch, estimate
+        the score and calculate the loss value.
+
+        Note that this step does not include sampler.
+        """
+        batch_size = target_signal.size(0)
+
+        if self.normalize_input:
```
Comment (racoiaws) [inline nemo/collections/asr/models/enhancement_models.py:468]: Didn't we normalize inputs in the `ASRAudioProcessor`? Or was that an unrelated change?

Comment (racoiaws) [reply nemo/collections/asr/models/enhancement_models.py:468]: Seems redundant

File: nemo/collections/asr/models/enhancement_models.py [hunk 10]
```diff
@@ -464,3 +199,380 @@ def list_available_models(cls) -> Optional[PretrainedModelInfo]:
...
+
+        return loss
+
+    # PTL-specific methods
+    def training_step(self, batch, batch_idx):
+        input_signal, input_length, target_signal, target_length = batch
+
+        # For consistency, the model uses multi-channel format, even if the channel dimension is 1
+        if input_signal.ndim == 2:
+            input_signal = input_signal.unsqueeze(1)
```
Comment (racoiaws) [inline nemo/collections/asr/models/enhancement_models.py:517]: Nitpick: would be really nice to have this as `einops.rearrange` call, this would document dimensions.

Same for other `squeeze`/`unsqueeze` calls inside methods like `training_step`

File: nemo/collections/asr/data/audio_to_audio.py (hunk 3 above)
Comment (racoiaws) [reply nemo/collections/asr/data/audio_to_audio.py:334]: All makes sense as long as the user understands this is about paired data

Thanks for adding the comment

File: nemo/collections/asr/parts/submodules/diffusion.py [hunk 11]
```diff
@@ -0,0 +1,1310 @@
...
+    def forward(self, input, input_length=None, condition=None):
+        # Stack real and imaginary components
+        B, C_in, D, T = input.shape
+
+        if C_in != self.in_channels:
+            raise RuntimeError(f'Unexpected input channel size {C_in}, expected {self.in_channels}')
+
+        # Stack real and imaginary parts
+        input_real_imag = [torch.cat([input[:, [c], ...].real, input[:, [c], ...].imag], dim=1) for c in range(C_in)]
+        input = torch.cat(input_real_imag, dim=1)
```
[the remark this answers is not in the document]
Comment (racoiaws) [reply nemo/collections/asr/parts/submodules/diffusion.py]: Perfect, thanks

File: nemo/collections/asr/parts/submodules/diffusion.py (hunk 8 above)
Comment (racoiaws) [reply nemo/collections/asr/parts/submodules/diffusion.py:1037]: > adding a new predictor/corrector would include actually implementing it and adding the option to the initializer of PredictorCorrectorSampler.

That's exactly my point - user has to modify NeMo instead of using it as a library/framework

I suggest to keep this thread open for the next reviewer to chime in

File: nemo/collections/asr/models/enhancement_models.py (hunk 9 above)
Comment (racoiaws) [reply nemo/collections/asr/models/enhancement_models.py:469]: Ok, makes sense for a specific model to have its own normalization (even if it's simply gain towards 0 dBFS)

**Review by racoiaws — COMMENTED** [review]
LGTM, little things I've noticed were all addressed, thanks @anteju 

I've already seen and worked with this code, so it would be good to get another review

File: examples/audio_tasks/conf/predictive.yaml [hunk 12]
```diff
@@ -0,0 +1,132 @@
+name: "predictive_model"
+
+model_type: predictive
```
Comment (titu1994) [inline examples/audio_tasks/conf/predictive.yaml:3]: Why is it outside model: scope ?

File: examples/audio_tasks/conf/predictive.yaml [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+  sample_rate: 16000
+  skip_nan_grad: false
+  num_outputs: 1
+  normalize_input: true # normalize the input signal to 0dBFS
+
+  train_ds:
+    manifest_filepath: ???
+    input_key: noisy_filepath
+    target_key: clean_filepath
+    audio_duration: 2.04 # 256 frames
```
Comment (titu1994) [inline examples/audio_tasks/conf/predictive.yaml:15]: 2.04 is 256 frames ?

File: examples/audio_tasks/conf/predictive.yaml [hunk 14]
```diff
@@ -0,0 +1,132 @@
...
+    input_key: noisy_filepath
+    target_key: clean_filepath
+    batch_size: 8
+    shuffle: false
+    num_workers: 4
+    pin_memory: true
+
+  encoder:
+    _target_: nemo.collections.asr.modules.audio_preprocessing.AudioToSpectrogram
+    fft_length: 510
```
Comment (titu1994) [inline examples/audio_tasks/conf/predictive.yaml:34]: Seems like an odd number, can you add some comment as to why it's picked ?

File: nemo/collections/asr/data/audio_to_audio.py [hunk 15]
```diff
@@ -812,6 +830,8 @@ class AudioToTargetDataset(BaseAudioDataset):
                                 If `None`, all channels will be loaded.
         target_channel_selector: Optional, select subset of channels from each input audio file.
                                  If `None`, all channels will be loaded.
+        normalization_signal: Normalize audio signals with a scale that ensures the normalization signal is in range [-1, 1].
+                              All audio signals are scaled by the same factor. No normalization if set to `None`.
```
Comment (titu1994) [inline nemo/collections/asr/data/audio_to_audio.py:834]: What are the str values supported ?

File: nemo/collections/asr/losses/audio_losses.py [hunk 16]
```diff
@@ -410,3 +416,161 @@ def forward(
...
+
+        # weight buffer
+        if weight is not None:
+            if any([w <= 0 for w in weight]):
+                raise ValueError(f'Weight must be positive! Current value: {weight}')
+            elif not np.isclose(sum(weight), 1, atol=1e-6):
+                raise ValueError(f'Weight should add to one, current weight: {weight}')
+            weight = torch.tensor(weight).reshape(1, -1)
+            logging.info(f'Channel weight set to %s', weight)
+        self.register_buffer('weight', weight)
```
Comment (titu1994) [inline nemo/collections/asr/losses/audio_losses.py:496]: Can you explicitly add persist=False ?

File: nemo/collections/asr/models/audio_to_audio_model.py [hunk 17]
```diff
@@ -158,23 +161,265 @@ def multi_validation_epoch_end(self, outputs, dataloader_idx: int = 0):
...
+        is_concat = config.get('is_concat', False)
+        if is_concat:
+            raise NotImplementedError('Concat not implemented')
+
+        # TODO: Consider moving `inject` from `audio_to_text_dataset` to a utility module?
+        # Automatically inject args from model config to dataloader config
+        inject_dataloader_value_from_model_config(self.cfg, config, key='sample_rate')
+
+        # Instantiate tarred dataset loader or normal dataset loader
+        if config.get('is_tarred', False):
```
Comment (titu1994) [inline nemo/collections/asr/models/audio_to_audio_model.py:175]: Any plans to add it ?

File: nemo/collections/asr/models/enhancement_models.py [hunk 18]
```diff
@@ -403,9 +139,9 @@ def training_step(self, batch, batch_idx):
         # Expand channel dimension, if necessary
         # For consistency, the model uses multi-channel format, even if the channel dimension is 1
         if input_signal.ndim == 2:
-            input_signal = input_signal.unsqueeze(1)
+            input_signal = einops.rearrange(input_signal, 'B T -> B 1 T')
```
Comment (titu1994) [inline nemo/collections/asr/models/enhancement_models.py:142]: Just a note, I dunno if you ever plan to export this to onnx but it doesn't support einops

File: nemo/collections/asr/models/enhancement_models.py [hunk 19]
```diff
@@ -464,3 +200,380 @@ def list_available_models(cls) -> Optional[PretrainedModelInfo]:
...
+
+        # SDE
+        self.sde = self.from_config_dict(self._cfg.sde)
+
+        # Sampler
+        if 'sde' in self._cfg.sampler:
+            raise ValueError('SDE should be defined in the model config, not in the sampler config')
+        if 'score_estimator' in self._cfg.sampler:
+            raise ValueError('Score estimator should be defined in the model config, not in the sampler config')
+
```
Comment (titu1994) [inline nemo/collections/asr/models/enhancement_models.py:367]: Hmm we're using custom NeMo samplers ? Any plans for diffusers compatibility ?

File: nemo/collections/asr/models/enhancement_models.py [hunk 20]
```diff
@@ -464,3 +200,380 @@ def list_available_models(cls) -> Optional[PretrainedModelInfo]:
...
+                sequences.
+
+        Returns:
+            Output signal `output` in the time domain and the length of the output signal `output_length`.
+        """
+        batch_length = input_signal.size(-1)
+
+        if self.normalize_input:
+            # max for each example in the batch
+            norm_scale = torch.amax(input_signal.abs(), dim=(-1, -2), keepdim=True)
```
Comment (titu1994) [inline nemo/collections/asr/models/enhancement_models.py:430]: Sidenote, take a look at https://arxiv.org/abs/2312.02696

File: nemo/collections/asr/modules/audio_modules.py [hunk 21]
```diff
@@ -1683,3 +1686,57 @@ def forward(self, mixture: torch.Tensor, estimate: torch.Tensor) -> torch.Tensor
...
+
+
+class SpectrogramEstimator(NeuralModule):
+    """Estimate spectral coefficients from the complex-valued input signal using a neural model.
+
+    This is a wrapper around a neural model which takes complex-valued spectral
+    coefficients as input and returns the estimated complex-valued spectral coefficients.
+    """
+
+    def __init__(self, *, model: str, **kwargs):
```
Comment (titu1994) [inline nemo/collections/asr/modules/audio_modules.py:1698]: Can we avoid * and ** kwargs

File: nemo/collections/asr/parts/submodules/diffusion.py [hunk 22]
```diff
@@ -0,0 +1,1310 @@
...
+__all__ = [
+    'OrnsteinUhlenbeckVarianceExplodingSDE',
+    'SpectrogramNoiseConditionalScoreNetworkPlusPlus',
+    'NoiseConditionalScoreNetworkPlusPlus',
+    'PredictorCorrectorSampler',
+]
+
+
+class StochasticDifferentialEquation(NeuralModule, ABC):
+    """Base class for stochastic differential equations.
```
Comment (titu1994) [inline nemo/collections/asr/parts/submodules/diffusion.py:40]: Shouldn't it be SDESolver rather than just at SDE

File: nemo/collections/asr/parts/submodules/diffusion.py [hunk 23]
```diff
@@ -0,0 +1,1310 @@
...
+
+        return drift, diffusion
+
+    @abstractmethod
+    def copy(self):
+        """Create a copy of this SDE.
+        """
+        pass
+
+    def __str__(self):
```
Comment (titu1994) [inline nemo/collections/asr/parts/submodules/diffusion.py:163]: Maybe override repr instead of str

File: nemo/collections/asr/parts/submodules/diffusion.py [hunk 24]
```diff
@@ -0,0 +1,1310 @@
...
+        return {
+            "output": NeuralType(('B', 'D'), VoidType()),
+        }
+
+    def forward(self, input):
+        x_proj = input[:, None] * self.W[None, :] * 2 * math.pi
+        return torch.cat([torch.sin(x_proj), torch.cos(x_proj)], dim=-1)
+
+
+class ResnetBlockBigGANPlusPlus(torch.nn.Module):
```
Comment (titu1994) [inline nemo/collections/asr/parts/submodules/diffusion.py:871]: Why is a non reset in t sde solver file?

File: nemo/collections/asr/parts/submodules/diffusion.py (hunk 8 above)
Comment (titu1994) [reply nemo/collections/asr/parts/submodules/diffusion.py:1037]: I like the hierarchy, don't remove base classes. Agreed that one can replace the whole class from config. 

Another consideration can be to use a registry of supported classes along with above class, but it seems redundant. Registry is more useful when you need an indirection from name to object, here that's not the case. 

However, it's important to note that some samplers are hard coded for specific model architectures and not fully general (at least in vision diffusion world) so it might be useful to have some sort of subclass test to ensure compatible samplers are used. It can be done in another pr

File: nemo/collections/asr/parts/submodules/diffusion.py [hunk 25]
```diff
@@ -0,0 +1,1310 @@
...
+
+        if state_length is not None:
+            new_state = mask_sequence_tensor(new_state, state_length)
+            mean = mask_sequence_tensor(mean, state_length)
+
+        return new_state, mean
+
+
+class Corrector(NeuralModule, ABC):
+    """Corrector for the reverse process.
```
Comment (titu1994) [inline nemo/collections/asr/parts/submodules/diffusion.py:1202]: I wonder if it's better to move predictor and corrector modules to their own files, if you expect many variations of them in the future

**Review by titu1994 — COMMENTED** [review]
Shallow review, went through most of the code apart from the samplers. Overall looks pretty good, minor comments mostly

File: examples/audio_tasks/conf/predictive.yaml (hunk 12 above)
Comment (nithinraok) [reply examples/audio_tasks/conf/predictive.yaml:3]: is this addressed?

File: examples/audio_tasks/conf/score_based_generative.yaml [hunk 26]
```diff
@@ -0,0 +1,151 @@
+name: score_based_generative_model
+
+model_type: score_based
```
Comment (nithinraok) [inline examples/audio_tasks/conf/score_based_generative.yaml:3]: move to model scope here as well.

File: nemo/collections/asr/data/audio_to_audio.py [hunk 27]
```diff
@@ -130,13 +130,18 @@ class ASRAudioProcessor:
...
         random_offset: If `True`, offset will be randomized when loading a subsegment
                        from a file.
+        normalization_signal: Normalize all audio with a factor that ensures the normalization
+                    signal is in range [-1, 1]. All audio signals are scaled by the same factor.
+                    Default is `None`, corresponding to no normalization.
     """
 
     def __init__(
-        self, sample_rate: float, random_offset: bool,
+        self, sample_rate: float, random_offset: bool, normalization_signal: Optional[str] = None, eps: float = 1e-8,
```
Comment (nithinraok) [inline nemo/collections/asr/data/audio_to_audio.py:139]: normalization signal doc string is bit unclear, is input torch array?

File: examples/audio_tasks/conf/score_based_generative.yaml [hunk 28]
```diff
@@ -0,0 +1,151 @@
...
+  decoder:
+    _target_: nemo.collections.asr.modules.audio_preprocessing.SpectrogramToAudio
+    fft_length: ${model.encoder.fft_length} 
+    hop_length: ${model.encoder.hop_length}
+    magnitude_power: ${model.encoder.magnitude_power}
+    scale: ${model.encoder.scale}
+
+  estimator:
+    _target_: nemo.collections.asr.modules.audio_modules.SpectrogramEstimator
+    model: ncsn++
```
Comment (nithinraok) [inline examples/audio_tasks/conf/score_based_generative.yaml:50]: Add comment that only `ncsn++` is supported?

File: nemo/collections/asr/modules/audio_preprocessing.py [hunk 29]
```diff
@@ -776,6 +792,14 @@ def forward(
         with torch.cuda.amp.autocast(enabled=False):
             output = self.stft(input.float())
 
+            if self.magnitude_power != 1:
```
Comment (nithinraok) [inline nemo/collections/asr/modules/audio_preprocessing.py:795]: curious to understand: is this a learnable module? I doesn;t seem so, hence why not mention `no_grad` to forward here?

File: nemo/collections/asr/modules/audio_modules.py (hunk 21 above)
Comment (nithinraok) [reply nemo/collections/asr/modules/audio_modules.py:1698]: @anteju in case this was missed

File: nemo/collections/asr/parts/submodules/diffusion.py [hunk 30]
```diff
@@ -0,0 +1,1310 @@
+# Copyright (c) 2023, NVIDIA CORPORATION & AFFILIATES.  All rights reserved.
```
Comment (nithinraok) [inline nemo/collections/asr/parts/submodules/diffusion.py:1]:
```suggestion
# Copyright (c) 2024, NVIDIA CORPORATION & AFFILIATES.  All rights reserved.
```

**Review by nithinraok — COMMENTED** [review]
Didn’t delve into diffusion-based algorithms; Did a top-level review!

**Comment by nithinraok** [comment]
jenkins

**Comment by nithinraok** [comment]
jenkins

**Review by nithinraok — DISMISSED** [review]
LGTM
